Fix AST transformations of by-name parameters

smithjessk opened the pull request:
Closes #24.
- Stop canonicalization if parameters are by-name. This works even if the function has variable arguments
- Add tests to cover these cases. These tests are inspired by / copied from [Scala Async](https://github.com/scala/async/blob/4395afe0ab7c59f088607d0849abe48b0f83265d/src/test/scala/scala/async/run/anf/AnfTransformSpec.scala). 
- Correct the implementation of [a test](https://github.com/smithjessk/coroutines/blob/bd1c75aeebd2ee344171ad11ae6163204f3a0e04/src/test/scala/org/coroutines/async-await-tests.scala#L233) also from Scala Async

smithjessk committed:
Partially implement reflection test from Scala Async

smithjessk committed:
Merge branch 'master' of https://github.com/storm-enroute/coroutines into add-async-tests

smithjessk committed:
Copy method definitions into canonicalized AST

smithjessk committed:
Copy declarations and identifiers from method definitions into proper scope

smithjessk committed:
Merge branch 'master' of https://github.com/storm-enroute/coroutines into add-async-tests

smithjessk committed:
Remove checks that asserted equality between unit types

smithjessk committed:
Make some progress on properly transforming by-name parameters

However, this commit is not stable. There is an infinite loop when compiling `async-await-tests.scala`.

smithjessk committed:
Remove infinite looping introduced in the previous commit

smithjessk committed:
Merge master

smithjessk committed:
Re-add tests that were accidentally removed in previous commit

smithjessk committed:
Don't canonicalize nested methods

smithjessk committed:
Fix failing tests involving by-name parameters

smithjessk committed:
Remove failing and commented-out tests

smithjessk committed:
Remove tests and changes copied from add-async-tests

smithjessk/add-async-tests contains many async tests that do not pass. Some of these, along with some improvements to the code, were copied into this branch. This commit removes the changes and tests that have to do with method definitions.

smithjessk committed:
Document and simplify canonicalization of function application

smithjessk committed:
Clean up and document by-name parameter transformation

smithjessk committed:
Fix implementation of test from Scala Async

axel22 commented on project/Build.scala:
```diff
@@ -96,6 +96,7 @@ object CoroutinesBuild extends MechaRepoBuild {
       "org.scalatest" % "scalatest_2.11" % "2.2.6" % "test",
       "org.scala-lang.modules" %% "scala-parser-combinators" % "1.0.2",
       "org.scala-lang" % "scala-reflect" % "2.11.4",
+      "org.scala-lang" % "scala-compiler" % "2.11.4",
```
How come we need this dependency?

smithjessk replied:
I added it while experimenting with different solutions. Forgot to take it out.

Removed it in 1387ecb.

axel22 commented on src/main/scala/org/coroutines/AstCanonicalization.scala:
```diff
@@ -8,7 +8,6 @@ import scala.language.experimental.macros
 import scala.reflect.macros.whitebox.Context
 
 
-
```
Always 3 lines between the import section and the first top-level section.

smithjessk replied:
1bc8ded.

axel22 commented on src/main/scala/org/coroutines/AstCanonicalization.scala:
```diff
@@ -122,6 +121,48 @@ trait AstCanonicalization[C <: Context] {
     case q"$selector[..$tpts](...$paramss)" if tpts.length > 0 || paramss.length > 0 =>
       // application
+
+      // We first need to see which parameters, if any, are by-name. Those that are
```
It is enough to say:

```
Detect by-name parameters, and prevent canonicalizing them.
```

smithjessk replied:
1bc8ded.

axel22 commented on src/main/scala/org/coroutines/AstCanonicalization.scala:
```diff
@@ -122,6 +121,48 @@ trait AstCanonicalization[C <: Context] {
+      // should not be canonicalized.
+
+      // Maps to the parameter lists, saying whether or not they are by-name.
+      val byNameParams: immutable.Seq[immutable.Seq[Boolean]] = {
```
It should be sufficient to use just `Seq[Seq[Boolean]]`.

smithjessk replied:
Done in 1bc8ded.

axel22 commented on src/main/scala/org/coroutines/AstCanonicalization.scala:
```diff
@@ -122,6 +121,48 @@ trait AstCanonicalization[C <: Context] {
+
+      // Maps to the parameter lists, saying whether or not they are by-name.
+      val byNameParams: immutable.Seq[immutable.Seq[Boolean]] = {
+        // Check that `selector` has a non-trivial symbol. If it doesn't, we assume
```
What is a non-trivial symbol?

smithjessk replied:
You're right: non-trivial isn't the best way to describe what I meant.

Some methods threw errors when I called `asMethod` on them. These checks are my way of preventing those exceptions.

If I recall correctly, the problematic functions were lambdas.

axel22 commented on src/main/scala/org/coroutines/AstCanonicalization.scala:
```diff
@@ -122,6 +121,48 @@ trait AstCanonicalization[C <: Context] {
+      val byNameParams: immutable.Seq[immutable.Seq[Boolean]] = {
+        // Check that `selector` has a non-trivial symbol. If it doesn't, we assume
+        // that there were no by-name parameters.
+        if (selector.symbol != null && selector.symbol != NoSymbol) {
```
When is the `selector.symbol` equal to `null` or `NoSymbol`?

smithjessk replied:
Related to https://github.com/storm-enroute/coroutines/pull/27#discussion_r70301309. 

I think lambdas were problematic. There might have been more cases, though. I should have documented this more heavily at the time.

axel22 commented on src/main/scala/org/coroutines/AstCanonicalization.scala:
```diff
@@ -122,6 +121,48 @@ trait AstCanonicalization[C <: Context] {
+            noRepeatedParamsSeq
+          }
+        } else {
+          immutable.Seq.fill(1, paramss(0).length)(false)
```
Does this mean that if there is no symbol information, all the parameters are considered non-by-name?

axel22 replied:
Should it not be:

```
for (params <- paramss) yield
  for (p <- params) yield false
```

Otherwise, with the current code you'd end up having only the first parameter list.

smithjessk replied:
About the lack of symbol information: yes, that's what I intended. 

About the first parameter list issue: you're right. I think I assumed that, if there was no appropriate symbol information, there would only be one parameter list. 

Changed this in 1bc8ded.

axel22 commented on src/main/scala/org/coroutines/AstCanonicalization.scala:
```diff
@@ -122,6 +121,48 @@ trait AstCanonicalization[C <: Context] {
+        if (selector.symbol != null && selector.symbol != NoSymbol) {
+          val paramLists = selector.symbol.asMethod.paramLists
+
+          // This value does not contain repeated parameters.
```
Note that repeated by name parameters are not allowed in Scala.

Also, the `paramss` value might not contain parameters with default values.

axel22 commented on src/main/scala/org/coroutines/AstCanonicalization.scala:
```diff
@@ -122,6 +121,48 @@ trait AstCanonicalization[C <: Context] {
+        } else {
+          immutable.Seq.fill(1, paramss(0).length)(false)
+        }
+      }
```
You should not inspect the method symbol and match the parameter trees to the method type information to figure out which is which - you're reproducing complex typechecker logic by doing that, and not covering all the edge cases.

Instead, see this:

http://stackoverflow.com/questions/29757584/handling-by-name-parameters-in-scala-macro

The `ByNameParamClass` is a dummy class that denotes that a parameter is by-name. Unfortunately, this dummy class is only used to wrap the type of the by-name parameter, but it does not appear in any way in an actual application. This is a flaw in the reflection API, as it does not expose enough information.

We should have the following behavior:

1. If the lengths of the argument lists at the callsite is equal to the length of the parameter lists at the method definition, AND, the respective argument types exactly match respective parameter types, then use that information to figure out what is by-name and what is not. This already covers a majority of cases.

2. Otherwise, do not try to patch the argument lists, and immediately report an error - such a method application cannot be used inside a coroutine definition. Workaround for users is to introduce some sort of bridge method.

smithjessk replied:
Can do! After I write unit tests, I will change my implementation.

smithjessk replied:
Working on this change [here](https://github.com/smithjessk/coroutines/tree/wip/change-by-name-detection).

smithjessk replied:
I've encountered [a problem](http://stackoverflow.com/questions/38706027/reflection-incorrectly-matching-by-name-parameters) with distinguishing between by-name and repeated parameters. Do you have any input?

smithjessk replied:
Would the first condition disallow calling methods with repeated parameters?

axel22 commented on src/main/scala/org/coroutines/AstCanonicalization.scala:
```diff
@@ -130,10 +171,34 @@ trait AstCanonicalization[C <: Context] {
       }
       for (tpt <- tpts) disallowCoroutinesIn(tpt)
-      val (pdeclss, pidents) = paramss.map(_.map(canonicalize).unzip).unzip
+
```
Could you do just:

```
val (pdeclss, pidents) = {
  for ((params, bynames) <- paramss.zip(bynamess)) yield {
    for ((p, byname) <- params.zip(bynames)) yield {
      if (byname) (List(), p) else canonicalize(p)
    }
  }
  cs.unzip.unzip
}
```

smithjessk replied:
By `bynamess`, do you mean `byNameParams`? 

Also, where is `cs` defined?

axel22 replied:
By `bynamess`, I meant `byNameParams`, yes.
`cs` is defined in my snippet, in the second line.

axel22 commented on src/test/scala/org/coroutines/async-await-tests.scala:
```diff
@@ -309,25 +307,47 @@ class AsyncAwaitTest extends FunSuite with Matchers {
   test("nested await as bare expression") {
     val c = async(coroutine { () =>
-      await(Future(await(Future("")).isEmpty))
+      val emptyString = await(Future(""))
```
Why this change?

smithjessk replied:
`Future.apply` takes a by-name parameter, but coroutines can't be passed by-name. See #24.

I will add some unit tests to assert that this check happens.

axel22 commented on src/test/scala/org/coroutines/async-await-tests.scala:
```diff
@@ -309,25 +307,47 @@ class AsyncAwaitTest extends FunSuite with Matchers {
     val c = async(coroutine { () =>
       ()
-      await(Future(await(Future("")).isEmpty))
+      val emptyString = await(Future(""))
```
Same here?

smithjessk replied:
See #24.

axel22 commented:
Please also add an additional file to tests `by-name-tests.scala`, using the ScalaTest template from another test file, and add multiple unit tests for the functionality that you are implementing.

smithjessk committed:
Remove dependency on scala-compiler

smithjessk committed:
Clean up by-name modifications in AstCanonicalization

smithjessk commented:
Yes! Will do. 

This is a note to myself to add unit tests for [this functionality](https://github.com/storm-enroute/coroutines/pull/27/files#r70304469).

smithjessk committed:
Remove commented-out code

smithjessk committed:
Add tests to assert that coroutines cannot be passed as by-name parameters

Note that these tests currently fail.

smithjessk committed:
Add repeated parameters test to ByNameTest

smithjessk committed:
Add test of by-name parameter traversal

Assert that they are not lifted even when they surround a non by-name parameter.

smithjessk committed:
Add another by-name parameter test

axel22 commented:
Should I take another look?

smithjessk commented:
I still have to go over a few of your comments. In the meantime, could you look at [this](https://github.com/storm-enroute/coroutines/pull/27#discussion_r70312250)? Thanks!

smithjessk committed:
Pass coroutine applications by-name in ByNameTest

Previously, the parameters passed by-name were unapplied coroutines.

smithjessk committed:
Run NestedContextValidator on by-name parameters

smithjessk commented:
To-do list for by-name analysis:

- [ ] Use `ByNameParam` class as described [here](https://github.com/storm-enroute/coroutines/pull/27#discussion_r70165760)

axel22 commented:
Note: this branch has merge conflicts.

smithjessk commented:
Yep- I'm still working on it. Do you mind taking a look at this [comment](https://github.com/storm-enroute/coroutines/pull/27#discussion_r70312250)?
